### ISSUE TITLE
caja-dropbox 1.18 trayicon/indicator is unusable - does not have menu entries on Ubuntu 17.10

### ISSUE THREAD

**Issue by N0rbert** [mate-desktop/caja-dropbox#27]
#### Expected behaviour
Dropbox trayicon/indicator menu is opened and it is usable.

#### Actual behaviour
Dropbox trayicon/indicator menu is opened but unusable (has empty entries, see screenshot).

![caja-dropbox has empty menu in trayicon](https://bugs.launchpad.net/bugs/1751451/+attachment/5061827/+files/caja-dropbox-empty-menu.png)

#### Steps to reproduce the behaviour
1. Install Ubuntu 17.10
2. Install `ubuntu-mate-desktop`
3. Install `caja-dropbox`, login, launch
4. Try to open menu of Dropbox trayicon/indicator

#### MATE general version
1.18.0-1

#### Package version
1.18.0-0ubuntu2

#### Linux Distribution
Ubuntu 17.10

#### Link to downstream report of your Distribution
https://bugs.launchpad.net/bugs/1751451

#### Notes
First seen on [Ubuntu MATE Commuinity](https://ubuntu-mate.community/t/caja-dropbox-not-showing-popup-menu-after-starting-mate-17-10/15780).

**Comment by alexarnaud** [comment]
Same on Debian with caja-dropbox 1.20, see [this downstream report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=893163).

**Comment by flexiondotorg** [comment]
@sunweaver I fixed this bug in Debian and Ubuntu with my last update to caja-dropbox 1.20 in pkg-mate, so this closed in Debian.

@raveit65 @monsta Does this issue exist in other distros? If so, I can make a pull request here.

**Comment by alexarnaud** [comment]
Le 17/03/2018 à 10:53, Martin Wimpress a écrit :
> @sunweaver <https://github.com/sunweaver> I fixed this bug in Debian and Ubuntu with my last update to caja-dropbox 1.20 in pkg-mate, so this closed in Debian.

Thank you Martin for the fix :)!

Which version should have the fixed ? Is it already pushed on Debian ?

Best regards,
Alex.

**Comment by N0rbert** [comment]
Just tested 1.20.0-2 on Debian Buster, bug exists:
![mate on Buster](https://user-images.githubusercontent.com/309456/37554195-2dcb062a-29e6-11e8-96da-171dcaa414ad.png).

I have removed Ubuntu 17.10, but in 18.04 LTS menu is in place.

And please do not forget about [`caja-dropbox autostart` bug](https://github.com/mate-desktop/caja-dropbox/issues/26).

**Comment by raveit65** [comment]
@flexiondotorg 
here the symptoms are different with SNI support enabled.
with left-click on applet i see Qt-theme.
![dropbox-na-tray-left-click](https://user-images.githubusercontent.com/961604/37554429-901e35da-29d8-11e8-8f8e-4973d8c5513a.png)
with right-click on applet i see gtk-theme + the issue with submenu.
![dropbox-na-tray-right-click](https://user-images.githubusercontent.com/961604/37554435-d54e4636-29d8-11e8-9aff-26fd54faed8a.png)

I think a PR won't hurt for testing.

**Comment by flexiondotorg** [comment]
@alexarnaud @sunweaver The comment by @raveit65 above is one example of where SNI is not doing the right thing and why I choose to disable.

**Comment by lukefromdc** [comment]
Why on earth would Caja-Dropbox ever show a QT menu or QT theme anywhere? How is QT getting into this? Find that and we might find the bug. I don't use Dropbox (or Google drive etc) so don't have this installed, but for a GTK app to ever use a QT theme strikes me as seriously odd. We should not even be loading QT unless another app has loaded it, and to my knowledge GTK does not read QT themes at all

**Comment by raveit65** [comment]
This part of caja-dropbox comes from proprietary dropbox code itself, so we can't fix it directly.
We ship only the caja extension code and a binary for start dropbox.
One solution is that people also file out reports against dropbox.......

**Comment by raveit65** [comment]
Btw. i think the switched from gtk+ to Qt 2 or 3 years ago.

**Comment by lukefromdc** [comment]
I looked in some of the caja-dropbox soruce files, saw includes for GTK but not QT. Dependency checks in configure.ac are

```
CAJA_REQUIRED=1.17.1
GLIB_REQUIRED=2.50.0
```
again no QT in those

**Comment by lukefromdc** [comment]
Ahh-now I see what you are saying. The indicator itself comes from Dropbox and not from Caja? If so, without even the source we have no idea what it is doing. That is not something I can fix myself.

**Comment by raveit65** [comment]
> The indicator itself comes from Dropbox and not from Caja?

Yes, same for this folder in home dir after installing the dropbox code.
[rave@mother ~]$ ls  ~/.dropbox-dist/
dropboxd  dropbox-lnx.x86_64-46.3.59  VERSION

**Comment by lukefromdc** [comment]
I just found that this redraw issue came when the panel failed to render the arrows in high contrast and was then restarted. The early parts of the code loaded either the custom CSS or the theme but not both, so I tried loading first the theme, then the custom CSS with this:

```
            /*load theme settings*/
	GtkSettings * settings;
	settings = gtk_settings_get_default ();
	gchar *gtk_theme_name;
	g_object_get (settings, "gtk-theme-name", &gtk_theme_name, NULL);
	css_provider = gtk_css_provider_get_named (gtk_theme_name, NULL);

	g_free (gtk_theme_name);

        	/* set custom css for size*/
	if (set_size) {
		gtk_css_provider_load_from_data (css_provider,
                                        ".panel_toplevel_hide_button {\n"
                                        "min-height: 17px; \n"
                                        "}", -1, NULL);
        }

	gtk_style_context_add_provider (context, GTK_STYLE_PROVIDER (css_provider), GTK_STYLE_PROVIDER_PRIORITY_SETTINGS);
```

**Comment by lukefromdc** [comment]
Wrong page comment, deleting...

**Comment by alexarnaud** [comment]
Le 17/03/2018 à 11:54, Martin Wimpress a écrit :
> @alexarnaud <https://github.com/alexarnaud> @sunweaver <https://github.com/sunweaver> The comment by @raveit65 <https://github.com/raveit65> above is one example of where SNI is not doing the right thing and why I choose to disable.

I'm not sure it is related to SNI support here. Dropbox on my system doesn't use SNI.

@raveit65: is it the same with SNI disabled ?

Best regards,
Alex.

**Comment by raveit65** [comment]
With SNI disabled the menus use Qt Styles with left/right click and all menuitems displaying well.
@flexiondotorg 
What fix did you proposed in you comment?
https://github.com/mate-desktop/caja-dropbox/issues/27#issuecomment-373908072

**Comment by flexiondotorg** [comment]
I'll file a pull request, but the fix makes copy of the environment then coerces XDG_DESKTOP_SESSION to report "Unity" and execs `dropbox` by passing in the modified environment. This environment modification is only exposed to the dropbox process and works around and issue where the Indicator only works correctly on Unity. The same fix has been applied to Ubuntu for Unity users because the session name changed to "Unity7" recently and also for Budgie.

**Comment by monsta** [comment]
This downstream patch by @flexiondotorg does the wrong thing when there are no indicators **and** SNI is disabled.

```diff
--- a/caja-dropbox.in	2017-03-28 02:00:33.223411568 +0100
+++ b/caja-dropbox.in	2017-03-28 02:03:34.637407866 +0100
@@ -728,8 +728,13 @@
     if os.access(db_path, os.X_OK):
         f = open("/dev/null", "w")
         # we don't reap the child because we're gonna die anyway, let init do it
+
+        # Fix indicator icon and menu. (LP: #1559249)
+        new_env = os.environ.copy()
+        new_env['XDG_CURRENT_DESKTOP'] = 'Unity'
+
         a = subprocess.Popen([db_path], preexec_fn=os.setsid, cwd=os.path.expanduser("~"),
-                             stderr=sys.stderr, stdout=f, close_fds=True)
+                             stderr=sys.stderr, stdout=f, close_fds=True, env=new_env)
 
         # in seconds
         interval = 0.5
```

Here's how it looks in Mint 19.1:

![bad-menu](https://user-images.githubusercontent.com/5138986/50734182-5e8f2f80-11ac-11e9-995f-0afeab884f1d.png)

Yes, this white block is the "menu".

If I enable SNI, the menu is back to normal.

If I revert the mentioned patch, the menu is also back to normal, this time with or without SNI. Actually, dropbox doesn't seem to use SNI if that patch is reverted. I get the same menu on left and right click, which would be impossible with SNI (left click would do nothing).

NOTE: Mint 19.1 is based on Ubuntu 18.04 and uses MATE 1.20 from Ubuntu repos. I chose it for testing because it has no indicators by default. It was easier than trying to rip all the indicators from the default Ubuntu setup...
